src/backend/distributed/relay/relay_event_utility.c

```diff
@@ -536,6 +547,70 @@ AppendShardIdToConstraintName(AlterTableCmd *command, uint64 shardId)
 }
 
 
+static bool
+ExtendRowReferencesWalker(Node *node, RowRefWalkerState *state)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, IndexElem))
+	{
+		IndexElem *indexElem = (IndexElem *) node;
+
+		return raw_expression_tree_walker(indexElem->expr, ExtendRowReferencesWalker,
+										  state);
+	}
+	else if (IsA(node, ColumnRef))
+	{
+		ColumnRef *columnRef = (ColumnRef *) node;
+		ListCell *fieldsCell;
+
+		/*
+		 * Append the shardId to any ColumnRef String values that are
+		 * equal to the relationName.  These are actually ROW(relname)
+		 * references.
+		 */
+		foreach(fieldsCell, columnRef->fields)
+		{
+			Value *fieldValue = (Value *) lfirst(fieldsCell);
+
+			if (IsA(fieldValue, String))
+			{
+				char *columnName = strVal(fieldValue);
+
+				if (strncmp(columnName, state->relationName, NAMEDATALEN) == 0)
+				{
+					AppendShardIdToName(&columnName, state->shardId);
+					fieldsCell->data.ptr_value = makeString(columnName);
+				}
+			}
+		}
+	}
+
+	return raw_expression_tree_walker(node, ExtendRowReferencesWalker, (void *) state);
+}
+
+
+/*
+ * AppendShardIdToRowReferences finds ColumnRef nodes that directly reference
+ * a column with the same name as the relation and extends those names with the
+ * given shardId.
+ */
+static void
+AppendShardIdToRowReferences(IndexStmt *indexStmt, char *relationName, uint64 shardId)
+{
+	RowRefWalkerState state;
+
+	state.relationName = relationName;
+	state.shardId = shardId;
+
+	raw_expression_tree_walker((Node *) indexStmt->indexParams, ExtendRowReferencesWalker,
```

Comment:
On 2016-08-25 14:51:54 -0700, Jason Petersen wrote:
> Heh, I hadn't looked at my comment while I was reading code. Yes,
> `IndexStmt` is a raw parsetree, so `raw_expression_tree_walker` is
> correct, unless we somehow do parse analysis on it.

And parse analysis can't be done before the relname has been changed. We
could hack that up by implementing our own version of
transformIndexStmt() and adding an alias, though. But that's not pretty.